walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,153 @@
+    // Add the whole day to potential meeting times
+    potentialMeetingTimes.add(TimeRange.WHOLE_DAY);
+
+    Collection<String> mandatoryMeetingGuests = request.getAttendees();
+    Collection<String> optionalMeetingGuests = request.getOptionalAttendees();
+    long meetingDuration = request.getDuration();
+    Collection<TimeRange> finalPossibleTimes = new ArrayList<>();
+
+    ArrayList<TimeRange> unavailableTimesForOptionalGuests = new ArrayList<>();
+    
+    // Check if there are any meeting guests in the event guest list, and if there are, remove the block of time 
+    // for the event from the potential meeting times 
+    for (Event event: events) {
+      TimeRange eventTime = event.getWhen();
+      Set<String> eventGuests = event.getAttendees();
+      // if there are any optional guests attending this event, add the time to the list of unavailable times for optional guests 
+      for (String optionalMeetingGuest: optionalMeetingGuests) {
+        if (eventGuests.contains(optionalMeetingGuest)) {
+          unavailableTimesForOptionalGuests.add(eventTime);
+        }
+      }
+      for (String mandatoryMeetingGuest: mandatoryMeetingGuests) {
+        if (eventGuests.contains(mandatoryMeetingGuest)) {
+          fixSchedulingConflicts(potentialMeetingTimes, eventTime);
+        }
+      }
+    }
+    
+    // Make a copy of good meeting times we have at this point for mandatory guests as a backup, in case scheduling is impossible with optional guests
+    ArrayList<TimeRange> copyOfGoodMeetingTimes = new ArrayList<TimeRange>(potentialMeetingTimes);
+    // Make a hashmap of all eliminated times and frequency of elimination, to optimize scheduling with the most optional guests 
+    HashMap<TimeRange, Integer> eliminatedTimes = new HashMap<>();
+ 
+    for (TimeRange unavailableTime: unavailableTimesForOptionalGuests) {
+      // boolean to check if an unavailable time has already been deleted
+      boolean timeAccounted = false;
+      // create a list iterator for potential meeting times so it can be modified while iterating 
+      ListIterator<TimeRange> litr = potentialMeetingTimes.listIterator();
+      while (litr.hasNext()) {
+        TimeRange potentialTime = litr.next();
+        if (unavailableTime.overlaps(potentialTime)) {
+          litr.remove();
+          // put unavailable time in eliminated times hasmap and assign frequency of 1 first time it is deleted
+          eliminatedTimes.put(potentialTime, 1);
+          if (potentialTime.start() < unavailableTime.start()) {
+            TimeRange frontBlock = TimeRange.fromStartEnd(potentialTime.start(), unavailableTime.start(), false);
+            litr.add(frontBlock);
+          }
+          if (unavailableTime.end() < potentialTime.end()) {
+            TimeRange backBlock = TimeRange.fromStartEnd(unavailableTime.end(), potentialTime.end(), false);
+            litr.add(backBlock);
+          }
+        }
+        timeAccounted = true; 
```

Comment:
This is a stylistic choice, but since deciding to set this to `true` is only dependent on `if (unavailableTime.overlaps(potentialTime))` I'd move this line right after the `litr.remove()` call so the ideas are connected.